add tests for symbol levels

feerrenrut opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft. See https://github.com/nvaccess/nvda/wiki/Contributing
-->

### Link to issue number:
Related to issue #11779, #10855, and PR #12710

### Summary of the issue:
PR #12710 has been suffering from scope creep, several of the problems discussed pre-existed in NVDA.

With Symbol level all:
-  '👕' symbol ("t-shirt") spoken as 't dash shirt' when moving by word or character
- Text "don't" spoken as "don tick t" when moving by word
- Symbols in NVDA speech UI (EG "spelling error" in French "Faute d'orthographe") are processed as symbols. The example "Faute d'orthographe" becomes "Faute d apostrophe orthographe", equivalent to "spelling e tick rror" in English, except that "apostrophe" is a longer word than "tick".

### Description of how this pull request fixes the issue:
Note, this PR does not introduce user visible changes. It is intended to help developers reason about the behavior of NVDA.

- Test (to demonstrate) behavior with Symbol level All and None (as the boundary cases for behavior) when moving by word, line, character.
- Test (to demonstrate) behavior with Symbol Level All when speech UI contains a symbol. For this, the System Test Spy global plugin had to be modified to alter translation strings of NVDA at runtime.

### Testing strategy:
System tests.

### Known issues with pull request:
None

### Change log entries:
None

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

feerrenrut committed:
Add tests for move by line and char

"symbol level word" tests are disabled, expectations for these tests
match the default symbol level tests to make change in behavior more
obvious.

feerrenrut committed:
remove unused tests, test symbol level all and none

feerrenrut committed:
Consistant sym level None, all for tests

feerrenrut committed:
test symbol substitution in Speech UI

lukaszgo1 commented on tests/system/robot/notepadTests.py:
```diff
@@ -14,49 +17,289 @@
+	CHAR = "numpad3"
+	WORD = "numpad6"
+	LINE = "numpad9"
+	HOME = "control+home"
```
Since you're using review cursor for navigation it would be much clearer to use Shift+numpad 7 here. Currently this relies on the fact that review cursor follows caret and therefore it is moved to the top of the document and it is better to be explicit about the fact that you want to move review cursor rather than system caret.

feerrenrut committed:
Reset virtual cursor to home directly, not via caret

feerrenrut commented:
We may want to add a test for symbols in table headers, see https://github.com/nvaccess/nvda/pull/12710#issuecomment-1031277294

feerrenrut commented:
We may also want to add tests for selecting text by word, see https://github.com/nvaccess/nvda/issues/11779#issuecomment-882307932

seanbudd reviewed (dismissed):
LGTM, do we have issues for the other test cases you want to cover? Or should they go in this PR?

feerrenrut commented:
After looking over #11779 and #10855 again, I think we should extend the move by word test to explicitly demonstrate multiple symbols combined, ie two bar symbols: `||`

I'll also look at adding tests for the selection and table cases. The table case will need to be a Chrome test.

feerrenrut committed:
Add test for double symbol #10855

feerrenrut committed:
Add comments about expectations, todo to track bugs

feerrenrut committed:
rename notepad tests

These test symbols / pronunciation. Will need to introduce chrome
for table tests.

feerrenrut committed:
Add pronunciation tests when selecting characters

feerrenrut committed:
add table col header symbol tests

feerrenrut commented:
I have added the extra test scenarios.

seanbudd commented on tests/system/robot/symbolPronunciationTests.py:
```diff
@@ -535,3 +538,76 @@ def _doTest(
+
+def test_tableHeaders():
+	"""
+	Announce the correct line when placed at the end of a link at the end of a list item in a contenteditable
```
This is a fairly specific scenario. It would be good to have a "why" for why this behaviour is being tested, like the other tests.

feerrenrut replied:
Ah yep, I forgot to go back to this. The table headers is an interesting example of "contextual" / secondary content. I.E. the table header text is still sourced from content, but it is not the primary text that is being traversed by the caret / review cursor. It raises the question of what symbol level should be used for secondary content.

seanbudd reviewed (commented):
LGTM other than the following addition. Due to the size of the PR, I plan to give it another read over just to see if anything else needs clarification.

feerrenrut committed:
fix comments for table header test

lukaszgo1 commented on tests/system/robot/symbolPronunciationTests.py:
```diff
@@ -0,0 +1,640 @@
+
+	There should not be any "empty" words. E.G. a word made of two bar chars: ||
+	Something should be reported, even at symbol level None. Possible:
+	- The names of the characters: "bar bar", what if there are many (hundreds) symbols?
```
I'd say this is a solved problem since when there are more than three characters in a row we just report their count assuming that the character in question is spoken at the current symbol level. I see no reason why this should work differently when navigating by word.

feerrenrut replied:
What if they are different characters, eg `!@#$%^`

feerrenrut replied:
`@^*_` is a better example for notepad.exe because it is treated as a word.

lukaszgo1 replied:
I don't see what is the problem here. If we decide to report the words at the symbol level all then each symbol should be reported to remain consistent with reporting when user sets symbol level to all manually.

feerrenrut committed:
Use several varied symbols as a word

Demonstrate another silient word case, where multiple symbols make up
a word (in notepad.exe). Expected behaviour unclear.

seanbudd reviewed (approved):
Looks good
